File: python/design_pattern/factory_method/main_factory_method.py
# /// script
# requires-python = ">=3.10"
# dependencies = []
# ///
"""
Factory Method Pattern

The Factory Method pattern defines an interface for creating an object, but lets
subclasses decide which class to instantiate. It lets a class defer instantiation
to subclasses.

Key Components:
- Product: Defines the interface of objects the factory method creates
- ConcreteProduct: Implements the Product interface
- Creator: Declares the factory method that returns a Product object
- ConcreteCreator: Overrides factory method to return a ConcreteProduct instance
"""

from abc import ABC, abstractmethod
from datetime import datetime


# Product interface
class Document(ABC):
    """Abstract product representing a document."""

    def __init__(self):
        self.content: list[str] = []
        self.created_at = datetime.now()

    @abstractmethod
    def get_type(self) -> str:
        pass

    @abstractmethod
    def get_extension(self) -> str:
        pass

    def add_content(self, text: str) -> None:
        self.content.append(text)

    def get_content(self) -> str:
        return "\n".join(self.content)

    @abstractmethod
    def render(self) -> str:
        pass


# Concrete Products
class PDFDocument(Document):
    """Concrete product for PDF documents."""

    def get_type(self) -> str:
        return "PDF Document"

    def get_extension(self) -> str:
        return ".pdf"

    def render(self) -> str:
        header = "%PDF-1.4\n"
        body = self.get_content()
        return f"{header}[PDF Binary Content]\n{body}\n%%EOF"


class HTMLDocument(Document):
    """Concrete product for HTML documents."""

    def get_type(self) -> str:
        return "HTML Document"

    def get_extension(self) -> str:
        return ".html"

    def render(self) -> str:
        content = self.get_content().replace("\n", "<br>\n")
        return f"""<!DOCTYPE html>
<html>
<head><title>Document</title></head>
<body>
{content}
</body>
</html>"""


class MarkdownDocument(Document):
    """Concrete product for Markdown documents."""

    def get_type(self) -> str:
        return "Markdown Document"

    def get_extension(self) -> str:
        return ".md"

    def render(self) -> str:
        return f"# Document\n\n{self.get_content()}"


class PlainTextDocument(Document):
    """Concrete product for plain text documents."""

    def get_type(self) -> str:
        return "Plain Text Document"

    def get_extension(self) -> str:
        return ".txt"

    def render(self) -> str:
        return self.get_content()


# Creator (abstract class with factory method)
class DocumentCreator(ABC):
    """Abstract creator that declares the factory method."""

    @abstractmethod
    def create_document(self) -> Document:
        """Factory method - subclasses must implement this."""
        pass

    def new_document(self, content: str) -> Document:
        """Template method that uses the factory method."""
        doc = self.create_document()
        doc.add_content(content)
        return doc

    def generate_report(self, title: str, sections: list[str]) -> Document:
        """Higher-level operation using the factory method."""
        doc = self.create_document()
        doc.add_content(f"Report: {title}")
        doc.add_content("-" * 40)
        for i, section in enumerate(sections, 1):
            doc.add_content(f"{i}. {section}")
        doc.add_content("-" * 40)
        doc.add_content(f"Generated: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
        return doc


# Concrete Creators
class PDFCreator(DocumentCreator):
    """Concrete creator for PDF documents."""

    def create_document(self) -> Document:
        return PDFDocument()


class HTMLCreator(DocumentCreator):
    """Concrete creator for HTML documents."""

    def create_document(self) -> Document:
        return HTMLDocument()


class MarkdownCreator(DocumentCreator):
    """Concrete creator for Markdown documents."""

    def create_document(self) -> Document:
        return MarkdownDocument()


class PlainTextCreator(DocumentCreator):
    """Concrete creator for plain text documents."""

    def create_document(self) -> Document:
        return PlainTextDocument()


# Parameterized Factory Method (alternative approach)
class DocumentFactory:
    """A parameterized factory that creates documents based on type."""

    _creators: dict[str, type[Document]] = {
        "pdf": PDFDocument,
        "html": HTMLDocument,
        "markdown": MarkdownDocument,
        "text": PlainTextDocument,
    }

    @classmethod
    def register(cls, doc_type: str, creator: type[Document]) -> None:
        """Register a new document type."""
        cls._creators[doc_type.lower()] = creator

    @classmethod
    def create(cls, doc_type: str) -> Document:
        """Create a document of the specified type."""
        creator = cls._creators.get(doc_type.lower())
        if creator is None:
            raise ValueError(f"Unknown document type: {doc_type}")
        return creator()


def main() -> None:
    print("=" * 60)
    print("Factory Method Pattern - Document Creation Demo")
    print("=" * 60)

    # Demo 1: Using concrete creators
    print("\n--- Using Concrete Creators ---")
    creators: list[DocumentCreator] = [
        PDFCreator(),
        HTMLCreator(),
        MarkdownCreator(),
        PlainTextCreator(),
    ]

    for creator in creators:
        doc = creator.new_document("Hello, Factory Method!")
        print(f"\n{doc.get_type()} ({doc.get_extension()}):")
        print("-" * 40)
        print(doc.render()[:200] + "..." if len(doc.render()) > 200 else doc.render())

    # Demo 2: Using factory to generate reports
    print("\n\n--- Generating Reports ---")
    html_creator = HTMLCreator()
    report = html_creator.generate_report(
        "Quarterly Sales",
        ["Revenue increased by 15%", "New customers: 250", "Retention rate: 92%"],
    )
    print(f"\n{report.get_type()} Report:")
    print(report.render())

    # Demo 3: Using parameterized factory
    print("\n--- Using Parameterized Factory ---")
    for doc_type in ["pdf", "html", "markdown", "text"]:
        doc = DocumentFactory.create(doc_type)
        doc.add_content("Created via parameterized factory")
        print(f"Created: {doc.get_type()}")

    # Demo 4: Registering custom document type
    print("\n--- Registering Custom Document Type ---")

    class XMLDocument(Document):
        def get_type(self) -> str:
            return "XML Document"

        def get_extension(self) -> str:
            return ".xml"

        def render(self) -> str:
            return f'<?xml version="1.0"?>\n<document>{self.get_content()}</document>'

    DocumentFactory.register("xml", XMLDocument)
    xml_doc = DocumentFactory.create("xml")
    xml_doc.add_content("Custom XML content")
    print(f"Created: {xml_doc.get_type()}")
    print(xml_doc.render())

    print("\n" + "=" * 60)
    print("Benefits of Factory Method Pattern:")
    print("=" * 60)
    print("1. Decouples client code from concrete product classes")
    print("2. Single Responsibility: creation code in one place")
    print("3. Open/Closed: add new products without modifying client")
    print("4. Provides hooks for subclasses to extend")


if __name__ == "__main__":
    main()

File: java/design_pattern/factory_method/MainFactoryMethod.java
/**
 * Comprehensive demonstration of the Factory Method Pattern in Java
 *
 * The Factory Method pattern defines an interface for creating an object,
 * but lets subclasses decide which class to instantiate. It lets a class
 * defer instantiation to subclasses.
 */

import java.util.HashMap;
import java.util.Map;

// Product interface
interface Document {
    void open();
    void save();
    void close();
    String getType();
}

// Concrete Products
class PDFDocument implements Document {
    private String name;

    public PDFDocument(String name) {
        this.name = name;
        System.out.println("  [PDFDocument] Created PDF document: " + name);
    }

    @Override
    public void open() {
        System.out.println("  [PDFDocument] Opening PDF: " + name);
    }

    @Override
    public void save() {
        System.out.println("  [PDFDocument] Saving PDF with compression: " + name);
    }

    @Override
    public void close() {
        System.out.println("  [PDFDocument] Closing PDF: " + name);
    }

    @Override
    public String getType() {
        return "PDF";
    }
}

class WordDocument implements Document {
    private String name;

    public WordDocument(String name) {
        this.name = name;
        System.out.println("  [WordDocument] Created Word document: " + name);
    }

    @Override
    public void open() {
        System.out.println("  [WordDocument] Opening Word doc: " + name);
    }

    @Override
    public void save() {
        System.out.println("  [WordDocument] Saving Word doc with formatting: " + name);
    }

    @Override
    public void close() {
        System.out.println("  [WordDocument] Closing Word doc: " + name);
    }

    @Override
    public String getType() {
        return "Word";
    }
}

class ExcelDocument implements Document {
    private String name;

    public ExcelDocument(String name) {
        this.name = name;
        System.out.println("  [ExcelDocument] Created Excel spreadsheet: " + name);
    }

    @Override
    public void open() {
        System.out.println("  [ExcelDocument] Opening Excel: " + name);
    }

    @Override
    public void save() {
        System.out.println("  [ExcelDocument] Saving Excel with formulas: " + name);
    }

    @Override
    public void close() {
        System.out.println("  [ExcelDocument] Closing Excel: " + name);
    }

    @Override
    public String getType() {
        return "Excel";
    }
}

class TextDocument implements Document {
    private String name;

    public TextDocument(String name) {
        this.name = name;
        System.out.println("  [TextDocument] Created text file: " + name);
    }

    @Override
    public void open() {
        System.out.println("  [TextDocument] Opening text file: " + name);
    }

    @Override
    public void save() {
        System.out.println("  [TextDocument] Saving plain text: " + name);
    }

    @Override
    public void close() {
        System.out.println("  [TextDocument] Closing text file: " + name);
    }

    @Override
    public String getType() {
        return "Text";
    }
}

// Creator (Abstract class with factory method)
abstract class Application {
    // Factory method - subclasses must implement
    public abstract Document createDocument(String name);

    // Template method that uses the factory method
    public void newDocument(String name) {
        Document doc = createDocument(name);
        doc.open();
        System.out.println("  Document ready for editing...");
    }

    public void saveAndClose(String name) {
        Document doc = createDocument(name);
        doc.save();
        doc.close();
    }
}

// Concrete Creators
class PDFApplication extends Application {
    @Override
    public Document createDocument(String name) {
        return new PDFDocument(name);
    }
}

class WordApplication extends Application {
    @Override
    public Document createDocument(String name) {
        return new WordDocument(name);
    }
}

class ExcelApplication extends Application {
    @Override
    public Document createDocument(String name) {
        return new ExcelDocument(name);
    }
}

class TextEditorApplication extends Application {
    @Override
    public Document createDocument(String name) {
        return new TextDocument(name);
    }
}

// Parameterized Factory Method example
class DocumentFactory {
    public Document createDocument(String type, String name) {
        switch (type.toLowerCase()) {
            case "pdf":
                return new PDFDocument(name);
            case "word":
                return new WordDocument(name);
            case "excel":
                return new ExcelDocument(name);
            case "text":
                return new TextDocument(name);
            default:
                throw new IllegalArgumentException("Unknown document type: " + type);
        }
    }
}

// Registry-based Factory example
class DocumentRegistry {
    private Map<String, Class<? extends Document>> registry = new HashMap<>();

    public void registerDocument(String type, Class<? extends Document> documentClass) {
        registry.put(type.toLowerCase(), documentClass);
    }

    public Document createDocument(String type, String name) {
        Class<? extends Document> documentClass = registry.get(type.toLowerCase());
        if (documentClass == null) {
            throw new IllegalArgumentException("Unknown document type: " + type);
        }

        try {
            return documentClass.getConstructor(String.class).newInstance(name);
        } catch (Exception e) {
            throw new RuntimeException("Failed to create document", e);
        }
    }
}

// Another example: Notification Factory
interface Notification {
    void send(String message);
}

class EmailNotification implements Notification {
    @Override
    public void send(String message) {
        System.out.println("  [Email] Sending email: " + message);
    }
}

class SMSNotification implements Notification {
    @Override
    public void send(String message) {
        System.out.println("  [SMS] Sending SMS: " + message);
    }
}

class PushNotification implements Notification {
    @Override
    public void send(String message) {
        System.out.println("  [Push] Sending push notification: " + message);
    }
}

class SlackNotification implements Notification {
    @Override
    public void send(String message) {
        System.out.println("  [Slack] Sending Slack message: " + message);
    }
}

abstract class NotificationService {
    public abstract Notification createNotification();

    public void notifyUser(String message) {
        Notification notification = createNotification();
        notification.send(message);
    }
}

class EmailService extends NotificationService {
    @Override
    public Notification createNotification() {
        return new EmailNotification();
    }
}

class SMSService extends NotificationService {
    @Override
    public Notification createNotification() {
        return new SMSNotification();
    }
}

class PushService extends NotificationService {
    @Override
    public Notification createNotification() {
        return new PushNotification();
    }
}

class SlackService extends NotificationService {
    @Override
    public Notification createNotification() {
        return new SlackNotification();
    }
}

public class MainFactoryMethod {
    public static void main(String[] args) {
        System.out.println("=== Factory Method Pattern Demonstration ===\n");

        // Using concrete creators
        System.out.println("--- 1. Using Concrete Creator Classes ---");

        System.out.println("\nPDF Application:");
        Application pdfApp = new PDFApplication();
        pdfApp.newDocument("report.pdf");

        System.out.println("\nWord Application:");
        Application wordApp = new WordApplication();
        wordApp.newDocument("letter.docx");

        System.out.println("\nExcel Application:");
        Application excelApp = new ExcelApplication();
        excelApp.newDocument("budget.xlsx");

        System.out.println("\nText Editor:");
        Application textApp = new TextEditorApplication();
        textApp.newDocument("notes.txt");

        // Using parameterized factory
        System.out.println("\n--- 2. Parameterized Factory Method ---");
        DocumentFactory factory = new DocumentFactory();

        Document doc1 = factory.createDocument("pdf", "invoice.pdf");
        doc1.open();
        doc1.save();
        doc1.close();

        System.out.println();

        Document doc2 = factory.createDocument("excel", "data.xlsx");
        doc2.open();
        doc2.save();
        doc2.close();

        // Using registry-based factory
        System.out.println("\n--- 3. Registry-Based Factory ---");
        DocumentRegistry registry = new DocumentRegistry();
        registry.registerDocument("pdf", PDFDocument.class);
        registry.registerDocument("word", WordDocument.class);
        registry.registerDocument("excel", ExcelDocument.class);
        registry.registerDocument("text", TextDocument.class);

        Document regDoc = registry.createDocument("word", "contract.docx");
        regDoc.open();
        regDoc.save();
        regDoc.close();

        // Notification example
        System.out.println("\n--- 4. Notification Factory Example ---");

        NotificationService emailService = new EmailService();
        emailService.notifyUser("Your order has been shipped!");

        NotificationService smsService = new SMSService();
        smsService.notifyUser("Verification code: 123456");

        NotificationService pushService = new PushService();
        pushService.notifyUser("New message received");

        NotificationService slackService = new SlackService();
        slackService.notifyUser("Build completed successfully");

        System.out.println("\n=== Summary ===");
        System.out.println("Factory Method pattern benefits:");
        System.out.println("  - Avoids tight coupling between creator and concrete products");
        System.out.println("  - Single Responsibility: product creation code in one place");
        System.out.println("  - Open/Closed: introduce new products without breaking existing code");
        System.out.println("  - Provides hooks for subclasses to extend");
    }
}
